### ISSUE TITLE
Option to insert an image from a URL

### ISSUE THREAD

**Issue by jasongisstl** [quilljs/quill#893]
The Image button on the toolbar allows you to insert an image from your local machine.
It would be brilliant if this could be extended to also allow an image URL to be provided.

It is possible to drag & drop an image from the web, but this may not be obvious to all users. Also, in our particular case, we have custom drag & drop handling which prevents this from working in our applications.

**Version: 1.0.0-rc2**

P.S. Thank you for all your hard work on 1.0 - it really is fantastic.

**Comment by laosb** [comment]
Any progress on this?

**Comment by CoenWarmer** [comment]
Also interested in this :)

**Comment by drewjbartlett** [comment]
Would also love to have this!

**Comment by dennisdebel** [comment]
I tried to implement this in the form of a custom module, but my JS skills suck. Behaviour im looking for is to replace all urls containing images to 'inline' images (like the etherpad image plugin). I made  (read: copypasta) an example that sorta works here: http://codepen.io/anon/pen/PpYjxG
In the example, type a bit and Copy/Paste an image url in the editor field and it will appear as inline image. Only after that the ability to edit/type stops, which must be something obvious, but as I said..major noob...

**Comment by tannerjt** [comment]
Would love to see this as well.  As a side note, this was how images were handled in the early stages of QuillJS.  Any reason why this was removed?

**Comment by masza** [comment]
I don't understand why there is base64 upload images and no support for third-party image upload. This is VERY important!

**Comment by cuduy197** [comment]
Still no update :(

**Comment by McSneaky** [comment]
I got it working like this: 
[https://stackoverflow.com/questions/43857373/quill-editor-vuejs2-image-upload-base64-image-to-url](https://stackoverflow.com/questions/43857373/quill-editor-vuejs2-image-upload-base64-image-to-url)

**Comment by Matoo125** [comment]
I think here https://github.com/quilljs/quill/issues/863#issuecomment-240579430 it is written pretty well. All you need to do is to create custom handler

**Comment by vhq2016** [comment]
+6666

**Comment by oliveratgithub** [comment]
+1

**Comment by tedtoy** [comment]
I too would like to make public my desire for such a feature.

**Comment by rodolfobubb** [comment]
+1

**Comment by Vice93** [comment]
+1 how is this not a feature yet?

**Comment by davek1979** [comment]
Could someone from the team maybe shed some light on this please? This feature is really, REALLY needed.

**Comment by oweskeez** [comment]
+1 my team would love this feature.

**Comment by alexjustesen** [comment]
For what it's worth... would be really really nice to have...

**Comment by hackerbrief** [comment]
Keeping the fire alive for this request 🔥

**Comment by oliveratgithub** [comment]
I toss some bricks into that fire so it keeps burning longer

**Comment by repat** [comment]
We got it working by extracting the base64 encoded image, storing it on the hard drive and replacing the` src="data..."` code with `src="/folder/file..."` but quill doesn't display the picture when the HTML is loaded with `dangerouslyPasteHTML` the next time we open it, see https://github.com/quilljs/quill/issues/1449.

+1 would also like to see this feature built-in

**Comment by Gibigbig** [comment]
```
var toolbarOptions = [
            ['bold', 'italic', 'underline', 'strike'],        // toggled buttons
            ['blockquote', 'code-block'],

            [{ 'header': 1 }, { 'header': 2 }],               // custom button values
            [{ 'list': 'ordered' }, { 'list': 'bullet' }],
            [{ 'script': 'sub' }, { 'script': 'super' }],      // superscript/subscript
            [{ 'indent': '-1' }, { 'indent': '+1' }],          // outdent/indent
            [{ 'direction': 'rtl' }],                         // text direction

            [{ 'size': ['small', false, 'large', 'huge'] }],  // custom dropdown
            [{ 'header': [1, 2, 3, 4, 5, 6, false] }],

            [{ 'color': [] }, { 'background': [] }],          // dropdown with defaults from theme
            [{ 'font': [] }],
            [{ 'align': [] }],

            ['clean'],                                         // remove formatting button
            ['image']
        ];

        var quill = new Quill('#editor', {
            theme: 'snow',
            modules: {
                toolbar: {
                    container: toolbarOptions,
                    handlers: {
                        image: imageHandler
                    }
                }
            },
        });

        function imageHandler() {
            var range = this.quill.getSelection();
            var value = prompt('What is the image URL');
            if(value){
                this.quill.insertEmbed(range.index, 'image', value, Quill.sources.USER);
            }
        }

**Comment by jiangqiming** [comment]
> Any progress on this?

啥

**Comment by siberiadev** [comment]
Hey there! Any progress?

**Comment by repat** [comment]
@siberiadev in case you're working with laravel, for extracting the image you can use `extract_inline_img()` from my package here: https://packagist.org/packages/repat/laravel-helper

Or just have a look at the function [here](https://github.com/repat/laravel-helper/blob/master/src/html_helpers.php) and extract what you need.

In terms of displaying I haven't seen any progress.

**Comment by siberiadev** [comment]
Repat, thank you. I use Node.js for backend.

**Comment by repat** [comment]
PHP and Laravel (MVC, similar to express I think) is fairly eary to read so you could:

* Have a look at the PHP documentation e.g. https://www.php.net/manual/de/function.preg-match-all.php to understand how my function extracts the base64 encoded data
* With https://regex101.com you can translate the Regular Expressions (e.g. `/src=\"data:image\/([a-zA-Z]*);base64,([^\"]*)\"/` to from PCRE to EcmaScript
* Other functions like `str_finish` come from the official Laravel helpers: https://laravel.com/docs/5.8/helpers

=> Translate it into a js function

**Comment by Gibigbig** [comment]
I posted a solution above

**Comment by jtlimson** [comment]
+1

**Comment by jtlimson** [comment]
This works like a charm

> ```
> var toolbarOptions = [
>             ['bold', 'italic', 'underline', 'strike'],        // toggled buttons
>             ['blockquote', 'code-block'],
> 
>             [{ 'header': 1 }, { 'header': 2 }],               // custom button values
>             [{ 'list': 'ordered' }, { 'list': 'bullet' }],
>             [{ 'script': 'sub' }, { 'script': 'super' }],      // superscript/subscript
>             [{ 'indent': '-1' }, { 'indent': '+1' }],          // outdent/indent
>             [{ 'direction': 'rtl' }],                         // text direction
> 
>             [{ 'size': ['small', false, 'large', 'huge'] }],  // custom dropdown
>             [{ 'header': [1, 2, 3, 4, 5, 6, false] }],
> 
>             [{ 'color': [] }, { 'background': [] }],          // dropdown with defaults from theme
>             [{ 'font': [] }],
>             [{ 'align': [] }],
> 
>             ['clean'],                                         // remove formatting button
>             ['image']
>         ];
> 
>         var quill = new Quill('#editor', {
>             theme: 'snow',
>             modules: {
>                 toolbar: {
>                     container: toolbarOptions,
>                     handlers: {
>                         image: imageHandler
>                     }
>                 }
>             },
>         });
> 
>         function imageHandler() {
>             var range = this.quill.getSelection();
>             var value = prompt('What is the image URL');
>             if(value){
>                 this.quill.insertEmbed(range.index, 'image', value, Quill.sources.USER);
>             }
>         }
> ```

**Comment by hidezzcooo** [comment]
his.quill.in### ### ### 

> [`_****_`]([url]([url]([url]([url]([url](

- @ #url

))))))

**Comment by chrisbmoore** [comment]
For anyone coming here using ngx-quill:

```
import { QuillModules, defaultModules } from 'ngx-quill';

export class MyComponent {

    quillModules: QuillModules = {
    toolbar: {
      container: defaultModules.toolbar, 
      handlers: {
        image: imageHandler
      }
    }
  };

}

function imageHandler(this: any) {
  const tooltip = this.quill.theme.tooltip;
  tooltip.save = function (this: any) {
    const range = this.quill.getSelection(true);
    const value = this.textbox.value;
    if (value) {
      this.quill.insertEmbed(range.index, 'image', value, 'user');
    }
  }
  tooltip.edit('image');
}
```

`<quill-editor [modules]="quillModules"></quill-editor>`

**Comment by pburrows** [comment]
@chrisbmoore this works great for ngx-quill! However, it clobbers the tooltip saves for links and videos. Need to set the original save back when done. Instead of this:

> ```
> function imageHandler(this: any) {
>   const tooltip = this.quill.theme.tooltip;
>   tooltip.save = function (this: any) {
>     const range = this.quill.getSelection(true);
>     const value = this.textbox.value;
>     if (value) {
>       this.quill.insertEmbed(range.index, 'image', value, 'user');
>     }
>   }
>   tooltip.edit('image');
> }
> ```

do something like this:

```
imageHandler(this: any) {
    const tooltip = this.quill.theme.tooltip;
    const originalSave = tooltip.save;
    tooltip.save = function(this: any) {
      const range = this.quill.getSelection(true);
      const value = this.textbox.value;
      if (value) {
        this.quill.insertEmbed(range.index, 'image', value, 'user');
      }
      tooltip.save = originalSave;
    };
    tooltip.edit('image');
  }
```

**Comment by chrisbmoore** [comment]
@pburrows You're right, the original save function should be reset. Thanks!

**Comment by digilist** [comment]
@pburrows  @chrisbmoore If the tooltip is closed without clicking "Save" the original save function is not restored, as the save method is not called. Do you have any idea how to solve this? I did not find some callback that is called in that case.

**Comment by chrisbmoore** [comment]
@digilist Thanks for the correction! Yep your solution makes sense and I'll add it to my code.

**Comment by joeabn** [comment]
Any Update on this ?

**Comment by connor4312** [comment]
I tried to find a better way to do this... the tooltips in Quill are hidden quite effectively so that consumers can't extend on them cleanly.

**Comment by tranduclinh2067** [comment]
> I tried to find a better way to do this... the tooltips in Quill are hidden quite effectively so that consumers can't extend on them cleanly.
> 
> Anyway, one more bit that you'll need to customize the label is this CSS snippet:
> 
> ```css
> .ql-tooltip[data-mode=image]::before {
>   content: "Enter image URL:";
> }
> ```

It doesn't work

**Comment by Milutin-P** [comment]
Other solutions did not work for me :(
I'm using a `"react-quill": "^1.3.5",`
So I wrote it like this:

```javascript
function imageHandler() {
  const { tooltip } = this.quill.theme;
  const originalSave = tooltip.save;
  const originalHide = tooltip.hide;

  tooltip.save = () => {
    const range = this.quill.getSelection(true);
    const { value } = tooltip.textbox;
    if (isValidURL(value)) { //isValidURL is a validator
      this.quill.insertEmbed(range.index, 'image', value, Quill.sources.USER);
    } else {
      // error message handler
    }
  };

  // Called on hide and save.
  tooltip.hide = () => {
    tooltip.save = originalSave;
    tooltip.hide = originalHide;
    tooltip.hide();
  };

  tooltip.edit('image');
  tooltip.textbox.placeholder = 'Embed URL';
}
```


```javascript
const modules = {
  toolbar: {
    container: [
      [{ header: '1' }, { header: '2' }, { font: [] }],
      [{ size: [] }],
      ['bold', 'italic', 'underline', 'strike', 'blockquote'],
      [{ list: 'ordered' }, { list: 'bullet' }],
      ['link', 'image', 'video'],
    ],
    handlers: {
      image: imageHandler,
    },
  },
};
```

**Comment by smujaddid** [comment]
> ```
> var toolbarOptions = [
>             ['bold', 'italic', 'underline', 'strike'],        // toggled buttons
>             ['blockquote', 'code-block'],
> 
>             [{ 'header': 1 }, { 'header': 2 }],               // custom button values
>             [{ 'list': 'ordered' }, { 'list': 'bullet' }],
>             [{ 'script': 'sub' }, { 'script': 'super' }],      // superscript/subscript
>             [{ 'indent': '-1' }, { 'indent': '+1' }],          // outdent/indent
>             [{ 'direction': 'rtl' }],                         // text direction
> 
>             [{ 'size': ['small', false, 'large', 'huge'] }],  // custom dropdown
>             [{ 'header': [1, 2, 3, 4, 5, 6, false] }],
> 
>             [{ 'color': [] }, { 'background': [] }],          // dropdown with defaults from theme
>             [{ 'font': [] }],
>             [{ 'align': [] }],
> 
>             ['clean'],                                         // remove formatting button
>             ['image']
>         ];
> 
>         var quill = new Quill('#editor', {
>             theme: 'snow',
>             modules: {
>                 toolbar: {
>                     container: toolbarOptions,
>                     handlers: {
>                         image: imageHandler
>                     }
>                 }
>             },
>         });
> 
>         function imageHandler() {
>             var range = this.quill.getSelection();
>             var value = prompt('What is the image URL');
>             if(value){
>                 this.quill.insertEmbed(range.index, 'image', value, Quill.sources.USER);
>             }
>         }
> ```

Inserting images with `insertEmbed` for images with URL does work, but quill doesn't allow it to be deleted, like when I press backspace or delete key, the inserted image won't disappear. Any way to solve this?

**Comment by jaymathew** [comment]
Thanks @Milutin-P, your code helped in 1.3.5,  unfortunately something changed on 1.3.7 and now your code snippet doesn't work. I think the issue lies with trying to get the tooltip.  ` const { tooltip } = this.quill.theme` I'm not sure if something changed but was wondering if you could shed some light on this issue.  

thanks,
